x-pack/functionbeat/provider/aws/op_wait_cloud_formation.go

```diff
@@ -5,68 +5,141 @@
 
 	"github.com/aws/aws-sdk-go-v2/aws"
 	"github.com/aws/aws-sdk-go-v2/service/cloudformation"
+	"github.com/aws/aws-sdk-go-v2/service/cloudformation/cloudformationiface"
 
 	"github.com/elastic/beats/libbeat/logp"
 )
 
-var periodicCheck = 10 * time.Second
+var periodicCheck = 2 * time.Second
 
-type opCloudWaitCloudFormation struct {
-	log       *logp.Logger
-	svc       *cloudformation.CloudFormation
-	stackName string
+type checkStatusFunc = func(*cloudformation.StackStatus) (bool, error)
+
+type opWaitCloudFormation struct {
+	log         *logp.Logger
+	svc         cloudformationiface.CloudFormationAPI
+	checkStatus checkStatusFunc
 }
 
 func newOpWaitCloudFormation(
+	log *logp.Logger,
+	svc cloudformationiface.CloudFormationAPI,
+) *opWaitCloudFormation {
+	return &opWaitCloudFormation{
+		log:         log,
+		svc:         svc,
+		checkStatus: checkCreateStatus,
+	}
+}
+
+func newWaitDeleteCloudFormation(
 	log *logp.Logger,
 	cfg aws.Config,
-	stackName string,
-) *opCloudWaitCloudFormation {
-	return &opCloudWaitCloudFormation{
-		log:       log,
-		svc:       cloudformation.New(cfg),
-		stackName: stackName,
+) *opWaitCloudFormation {
+	return &opWaitCloudFormation{
+		log:         log,
+		svc:         cloudformation.New(cfg),
+		checkStatus: checkDeleteStatus,
 	}
 }
 
-func (o *opCloudWaitCloudFormation) Execute() error {
-	o.log.Debug("Waiting for cloudformation confirmation")
-	status, reason, err := queryStackStatus(o.svc, o.stackName)
+func (o *opWaitCloudFormation) Execute(ctx executionContext) error {
+	c, ok := ctx.(*stackContext)
+	if !ok {
+		return errWrongContext
+	}
+
+	if c.ID == nil {
+		return errMissingStackID
+	}
+
+	eventStackPoller := makeEventStackPoller(o.log, o.svc, periodicCheck, c)
+	eventStackPoller.Start()
+	defer eventStackPoller.Stop()
 
-	// List of States from the cloud formation API.
-	// https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_Stack.html
 	for {
-		o.log.Debugf(
-			"Retrieving information on stack '%s' from cloudformation, current status: %v",
-			o.stackName,
-			*status,
-		)
+		status, _, err := queryStackStatus(o.svc, c.ID)
+		if err != nil {
+			return err
+		}
 
+		completed, err := o.checkStatus(status)
 		if err != nil {
 			return err
 		}
 
-		switch *status {
-		case cloudformation.StackStatusUpdateComplete: // OK
+		if completed {
 			return nil
-		case cloudformation.StackStatusCreateComplete: // OK
-			return nil
-		case cloudformation.StackStatusCreateFailed:
-			return fmt.Errorf("failed to create the stack '%s', reason: %v", o.stackName, reason)
-		case cloudformation.StackStatusRollbackFailed:
-			return fmt.Errorf("failed to create and rollback the stack '%s', reason: %v", o.stackName, reason)
-		case cloudformation.StackStatusRollbackComplete:
-			return fmt.Errorf("failed to create the stack '%s', reason: %v", o.stackName, reason)
 		}
 
 		select {
 		case <-time.After(periodicCheck):
-			status, reason, err = queryStackStatus(o.svc, o.stackName)
 		}
```

Comment:
We need a select here? looks like `time.Sleep(periodicCheck)`.

You plan to add a done-channel for shutdown here?

Reply:
I've removed the select since I do not plan to add a done channel, we will get out of the loop on any completion state or on error and this will have effect of closing the other poller.